sources/api/shibaken/src/main.rs

```diff
@@ -0,0 +1,318 @@
+*/
+
+#![deny(rust_2018_idioms)]
+
+#[macro_use]
+extern crate log;
+
+use reqwest::blocking::Client;
+use serde::Serialize;
+use simplelog::{Config as LogConfig, LevelFilter, TermLogger, TerminalMode};
+use snafu::{ensure, ResultExt};
+use std::collections::HashMap;
+use std::{env, process};
+use url::Url;
+
+// Setting generators do not require dynamic socket paths at this moment.
+const DEFAULT_API_SOCKET: &str = "/run/api.sock";
+const API_METADATA_URI_BASE: &str = "/metadata/";
+
+// Instance Meta Data Service.
+const IMDS_PUBLIC_KEY_BASE_URL: &str = "http://169.254.169.254/latest/meta-data/public-keys/";
+const IMDS_SESSION_TOKEN_ENDPOINT: &str = "http://169.254.169.254/latest/api/token";
+
+#[derive(Serialize)]
+struct UserData {
+    ssh: Ssh,
+}
+
+#[derive(Serialize)]
+struct Ssh {
+    authorized_keys: Vec<String>,
+}
+
+/// Returns the value of a metadata key for a given data key, erroring if the value is not a
+/// string or is empty.
+async fn get_metadata(key: &str, meta: &str) -> Result<String> {
+    let uri = &format!("{}{}?keys={}", API_METADATA_URI_BASE, meta, key);
+    let method = "GET";
+    let (code, response_body) = apiclient::raw_request(DEFAULT_API_SOCKET, &uri, method, None)
+        .await
+        .context(error::APIRequest { method, uri })?;
+    ensure!(
+        code.is_success(),
+        error::Response {
+            method,
+            uri,
+            code,
+            response_body
+        }
+    );
+
+    // Metadata responses are of the form `{"data_key": METADATA}` so we pull out the value.
+    let mut response_map: HashMap<String, serde_json::Value> =
+        serde_json::from_str(&response_body).context(error::DeserializeJson)?;
+    let response_val = response_map
+        .remove(key)
+        .context(error::MissingMetadata { meta, key })?;
+
+    // Ensure it's a non-empty string
+    let response_str = response_val
+        .as_str()
+        .with_context(|| error::MetadataWrongType {
+            meta,
+            expected: "string",
+            value: response_val.to_string(),
+        })?;
+    let metadata = if !response_str.is_empty() {
+        response_str.to_string();
+    } else {
+        String::new();
+    };
+    Ok(metadata)
+}
+
+fn get_text_from_imds(client: &Client, uri: &str, session_token: &str) -> Result<String> {
+    client
+        .get(uri)
+        .header("X-aws-ec2-metadata-token", session_token)
+        .send()
+        .context(error::ImdsRequest { method: "GET", uri })?
+        .error_for_status()
+        .context(error::ImdsResponse { uri })?
+        .text()
+        .context(error::ImdsText { uri })
+}
+
+/// Returns a list of public keys.
+fn get_public_keys_from_imds() -> Result<Vec<String>> {
+    info!("Fetching IMDS session token");
+    let client = Client::new();
+    // Use IMDSv2 for accessing instance metadata.
+    let uri = IMDS_SESSION_TOKEN_ENDPOINT;
+    let imds_session_token = client
```

Comment:
Take a peek at `early-boot-config`'s handling of session tokens.  There, it is split into a function and the errors are a bit more descriptive.  

`sources/api/early-boot-config/src/provider/aws.rs`  < `fetch_imds_session_token()`